### PR TITLE
Clean up cputopology initialization global communication pattern

### PR DESCRIPTION
**Pull request by evan-charmworks**
- Replaces error-prone locking and improper scheduler driving with well-tested pattern from Isomalloc-Sync
- Replaces bespoke "empty reduction" with CmiBarrier
- Replaces Broadcast with NodeBroadcast and moves send outside of handler
- Replaces unnecessary privatization of Converse handler index variables with CmiAssignOnce

### CODE REVIEW

**Comment by evan-charmworks** [comment]
@trquinn Does this resolve the hang on init from #2850?

**Review by ericjbohm — DISMISSED** [review]
Commit comment should note motive for switching from locked Cpv to atomic.

**Comment by evan-charmworks** [comment]
> Commit comment should note motive for switching from locked Cpv to atomic.

Updated.

File: src/conv-core/cputopology.C [hunk 1]
```diff
@@ -198,6 +196,32 @@ static void printTopology(int numNodes)
     CmiPrintf("Charm++> Running on %d hosts\n", numNodes);
 }
 
+static std::atomic<char> cpuTopoSyncHandlerDone;
+#if CMK_SMP && !CMK_SMP_NO_COMMTHD
+extern void CommunicationServerThread(int sleepTime);
+static std::atomic<char> cpuTopoSyncCommThreadDone;
```
Comment (matthiasdiener) [inline src/conv-core/cputopology.C:202]: Is there a reason this is not a `std::atomic<bool>`? Should this be explicitly initialized? (same  question for cpuTopoSyncHandlerDone)

Comment (evan-charmworks) [reply src/conv-core/cputopology.C]: Done

**Review by matthiasdiener — APPROVED** [review]

**Comment by evan-charmworks** [comment]
Needs one more review.

**Review by ericjbohm — APPROVED** [review]

**Pull request merged**